### PR TITLE
Networkmediaserver

### PR DESCRIPTION
**Pull request by bzeller**
This patch implements a media server that is seperated in its own thread, communicating over a socket to the main thread. This could be extracted into its own executable running with lesser priviledges if required.

### CODE REVIEW

**Review by mlandres — CHANGES_REQUESTED** [review]
+ the change required to compile with protobuf15

File: zypp/zyppng/media/medianetwork.cc [hunk 1]
```diff
@@ -214,49 +516,161 @@ void MediaHandlerNetwork::handleRequestResult( const Download &req ) const
...
+  curlcred->setAuthType( availAuth );
 
   std::string prompt_msg = zypp::str::Format( _("Authentication required for '%s'") ) % _url.asString();
 
-  // set available authentication types from the exception
-  // might be needed in prompt
-  authCopy.setAuthType(availAuth);
+  // the auth data that was used in the request is older than whats in the Credential manager,
+  // so instead of asking the user for creds we first try to use the new ones
+  if ( res->last_auth_timestamp() > 0 && res->last_auth_timestamp() < authDataPtr->lastDatabaseUpdate() ) {
```
Comment (mlandres) [inline zypp/zyppng/media/medianetwork.cc:567]:
```
/home/ma/zypp/libzypp/zypp/zyppng/media/medianetwork.cc:567:69: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if ( res->last_auth_timestamp() > 0 && res->last_auth_timestamp() < authDataPtr->lastDatabaseUpdate() ) {
```
`last_auth_timestamp` should match `time_t`

File: zypp/zyppng/media/medianetwork.cc [hunk 2]
```diff
@@ -268,42 +682,92 @@ std::shared_ptr<Download> MediaHandlerNetwork::prepareRequest( Downloader &dlMan
...
+  });
+  if ( i == _requests.end() )
+    return nullptr;
+  return &(*i);
+}
+
+MediaHandlerNetwork::Request *MediaHandlerNetwork::findRequest( const int id ) const
+{
+  const auto &i = std::find_if( _requests.begin(), _requests.end(), [ &id ]( const auto &r ) {
+    return ( id == r.requestid() );
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/zyppng/media/medianetwork.cc:746]: Done

File: zypp/zyppng/media/medianetwork.cc (hunk 1 above)
Comment (bzeller) [reply zypp/zyppng/media/medianetwork.cc:567]: Done

**Review by mlandres — APPROVED** [review]

**Pull request merged**